runtime/timelib_wrapper.cpp

```diff
@@ -0,0 +1,95 @@
+#include "runtime/timelib_wrapper.h"
+
+#include <kphp/timelib/timelib.h>
+
+#include "common/containers/final_action.h"
+#include "common/smart_ptrs/singleton.h"
+
+namespace {
+timelib_tzinfo *get_timezone_info(const char *tz_name) {
+  int dummy_error_code = 0;
+  timelib_tzinfo *info = timelib_parse_tzfile(tz_name, timelib_builtin_db(), &dummy_error_code);
+  return info;
+}
+} // namespace
+
+struct TzinfoCache : private vk::not_copyable {
+  timelib_tzinfo *etc_gmt3_;
+  timelib_tzinfo *europe_moscow_;
+
+  void init() {
+    // if we'll need other timezones, they can be passed as a command-line
+    // flag to instruct the runtime to load extra tzinfo objects here
+
+    europe_moscow_ = get_timezone_info("Europe/Moscow");
+    php_assert(europe_moscow_ != nullptr);
+
+    etc_gmt3_ = get_timezone_info("Etc/GMT-3");
+    php_assert(etc_gmt3_ != nullptr);
+  }
+
+  timelib_tzinfo *get_tzinfo(const char *tz_name) {
+    if (strcmp("Europe/Moscow", tz_name) == 0) {
+      return europe_moscow_;
+    }
+    if (strcmp("Etc/GMT-3", tz_name) == 0) {
+      return etc_gmt3_;
+    }
+    return nullptr;
+  }
+};
+
+void php_timelib::init() {
+  vk::singleton<TzinfoCache>::get().init();
+}
+
+int php_timelib::days_in_month(int64_t m, int64_t y) {
+  return timelib_days_in_month(y, m);
+}
+
+bool php_timelib::valid_date(int64_t m, int64_t d, int64_t y) {
+  return y >= 1 && y <= 32767 && timelib_valid_date(y, m, d);
+}
+
+std::pair<int64_t, bool> php_timelib::strtotime(const string &tz_name, const string &times, int64_t preset_ts) {
+  if (times.empty()) {
+    return {0, false};
+  }
+
+  timelib_tzinfo *tzi = vk::singleton<TzinfoCache>::get().get_tzinfo(tz_name.c_str());
+  if (tzi == nullptr) {
+    return {0, false};
+  }
+
+  bool use_heap_memory = (dl::get_script_memory_stats().memory_limit == 0);
+  auto malloc_replacement_guard = make_malloc_replacement_with_script_allocator(!use_heap_memory);
+
+  timelib_time *now = timelib_time_ctor();
+  auto now_deleter = vk::finally([now]() { timelib_time_dtor(now); });
+
+  // timelib destructors do timelib_free which is a macro that expands to free();
+  // so we probably don't need a CriticalSectionGuard here
```

Comment:
timelib_free call is not the main issue. The main issue is an invalid internal library state. If you sure, that timelib doesn't use any static/global vars inside (which can be spoiled with signal interruption), then we don't need a CriticalSectionGuard here, otherwise we must use the guard.

And if you add the guard, the second question is: Is timelib ready for nullptrs from malloc family calls?

Reply:
I updated the comment.

- timelib can't handle nullptr from malloc gracefully
- timelib is thread-safe

So it looks like CriticalSectionGuard is not needed here.